query.go

```diff
@@ -575,8 +575,11 @@ func marshal(q interface{}, wrapper string, fields Fields) (string, error) { //n
 
 	// Check to see if this type has already been built.
 	if cachedOperation, hit := cache.Load(rt); hit {
-		// Cache hit, use the tree that was already built.
-		operation = cachedOperation.(*field)
+		// Cache hit, copy by value the tree that was already built.
+		// A copy is needed because the top level declaration name (primitive type)
+		// of the operation is re-assigned.
+		opReplica := *cachedOperation.(*field)
+		operation = &opReplica
```

Comment:
I am afraid we need deeper refactoring.
Current "buggy" implementation modifies operation data (`Decl`) stored in cache after being retrieved from cache. So supposedly next time it is retrieved from cache it has new pieces.

This implementation modifies data on a copy of cached entry so that change is not visible to the next caller that hits the cache.

And maybe it is fine - the `Decl` part should be custom to each cycle. In that case we should probably removed it from cache entry and have it as local var in this function.

But if it is integral part of cache entry w'll probably need larger sync scope.

Reply:
You are right. The updated field in `Declaration` is definitely not integral part of the cache entry. Therefore, next retrievals don't need to be aware of it.

Reply:
Y'all are right on the money with your analysis as far as I can tell. I'll approve this in the interim though